### ISSUE TITLE
Docs are unusable without a mouse with JavaScript enabled 

### ISSUE THREAD

**Issue by SleeplessByte** [zzak/sdoc#158]
Cross-post: https://github.com/rails/rails/issues/41813

In general this results in the website not being usable by those who rely on screen readers, keyboard navigation, are visually impaired, or use other tools expecting (some form) of [WCAG](https://www.w3.org/WAI/standards-guidelines/wcag/) compliance. 

There are 3 quick wins here:

1. Do not override TAB, Space, Enter or SHIFT + TAB
2. Do not override system shortcuts (Windows, Mac, and Linux), such as CTRL / CMD + C, X, V
3. Show focus outline on links. If there is worry about the looks, a focus ring using `border` is fine too, as is [this solution](https://medium.com/hackernoon/removing-that-ugly-focus-ring-and-keeping-it-too-6c8727fefcd2).

Further wins:

1. Add skip navigation link
2. Add possibility to turn off shortcuts
3. Add possibility to remap shortcuts

It is important that we developers try to be inclusive, because being inclusive in tooling, especially when it's such a vital thing as documentation, allow those that need it most to contribute and be part of our (development) community too, and thus get their input on how we all can do better.

**Comment by p8** [comment]
Hey Derk-Jan, let's fix this.

There is a PR for removing the ctrl-C shortcut: https://github.com/zzak/sdoc/pull/155

Both TAB and Enter load the currently selected sidebar item into the body.
Removing the TAB override seem like a quick win, so I created a PR: https://github.com/zzak/sdoc/pull/159
I'm not sure what the expected behaviour would be for TAB. Should the next item be selected in the side panel?

**Comment by p8** [comment]
In 9eb5e92e70583bf0 a `tab-index="-1"` was added to the search input.
Is that recommended or should that be changed as well?
cc @MikeRogers0

**Comment by SleeplessByte** [comment]
It depends who we ask ;) 

In general tab will move focus to the next _interactive element_. When a component / widget is an interactive element, it should have a *single tab stop*. What does this mean in practice? If the menus sidebar is considered a widget, then it should have a single tab stop. 

In my opinion, the entire sidebar *is* a widget: it's a `search` widget with the list of suggestions. When the search field is empty, it shows _all_ suggestions. That makes the entire sidebar a search widget.

![Showing the autocomplete search widget on the rails docs](https://user-images.githubusercontent.com/1964376/113347471-804cd200-9335-11eb-8212-b05edd339737.png)

The widget _in this form_ doesn't exist in the WAI-ARIA Authoring Practices (version 1.1). It is, as far as I can tell a combination of:

- autosuggest
- combobox + tree view
- combobox + list view

Normally, autosuggest has a single input field and a popup/dropdown. In this case, the dropdown the sidebar (that's fine!). There are a LOT of rules how to implement these widgets compliantly, which may not be possible to do right away (thinking about time effort), but here are a few resources that should help:

- https://www.accessibility-developer-guide.com/examples/widgets/autosuggest/
- https://www.w3.org/TR/wai-aria-practices/#combobox
- https://www.w3.org/TR/wai-aria-practices/#TreeView (when not searching)

https://devdocs.io/ has an incomplete but "acceptable" implementation. Here are the things I recommend based on experience and knowing the practices.

1. The search box should not have `tabindex=-1`. It should not be hidden solely using `visiblity: hidden`, should use `hidden` or `aria-hidden="true"`. That will stop screen-readers from considering it. 
2. The entire sidebar is a single tab stop. So when I'm _in the search box_, and then press `tab`, it should go to the first focusable element on the right side.
3. Arrow up from the search field (searching or not searching) brings me to the _last_ item.
4. Arrow down from the search field (searching or not searching) brings me to the _first_ item.
5. Left and right arrow collapse / expand items (this is good, don't change).
6. Shift-tab from the first focusable element on the right / content should focus the search box (this will happen automatically if you remove `tabindex="-1"`)

Where/when is the search hidden? In most of my usage I always saw the search box :)

**Comment by p8** [comment]
Great response ❤️

**Comment by SleeplessByte** [comment]
Shout out when you / someone wants to do more. I am happy to walk whomever through what I know from the docs, the pitfalls, and/or make decisions on what is probably "Good Enough™".

For example: you'd probably be interested in marking up the sidebar as a proper search field, with a autocompletion list, announce results summary if they change as you type, implement a few shortcuts, such as "speed navigation" (per page / per 10 items up and down). These can all be incremental additions, in my opinion :)

**Comment by MikeRogers0** [comment]
I'm super onboard to help improve the accessibility & overall experience of the docs :)

One quick question to help me out massively:

1. What software are you using? I was testing locally using VoiceOver on iOS, but I'm always keen to duplicate your experience as much as I can :)

> Further wins: Add skip navigation link

I'll aim to get a PR in for this next :D

**Comment by SleeplessByte** [comment]
> What software are you using? I was testing locally using VoiceOver on iOS, but I'm always keen to duplicate your experience as much as I can :)

I'm usually on Windows (even though I also have a MacBook. I dislike VoiceOver (and lack of choice on macs), by default on [Chrome 89+](https://www.google.com/chrome/). I also have (and use) [Firefox 87+](https://www.mozilla.org/en-US/firefox/new/). [NVDA](https://www.nvaccess.org/download/) is my go-to screen reader, sometimes I use [JAWS](https://www.freedomscientific.com/products/software/jaws/).  I'm not a power user of any of these tools as I don't *require* or even use them during most days, but I do have access to people who absolutely require them, should you need it.

I use [Night Light](https://support.microsoft.com/en-us/windows/set-your-display-for-night-time-in-windows-10-18fe903a-e0a1-8326-4c68-fd23d7aaf136) (by windows), [Night Shift](https://support.apple.com/en-us/HT207513) (Mac), and [f.lux](https://justgetflux.com/). The last one is my go-to, but I'm trying out windows' tool so I can recommend it to people, if it's "Good Enough". Why does this matter? It changes colours, so I'm often someone that can tell that some colour combinations won't work for those how have some form of reduced-contrast impairment. When I do need to test some of this, I often use the built-in [Chrome Developers "Emulate vision deficiencies" tool](https://twitter.com/mathias/status/1237393102635012101).

As for further tools, [WAVE](https://wave.webaim.org/) is really decent and comes with browser extensions. Recently started trying out [IBM's tool](https://www.ibm.com/able/). Is very very nice thusfar.

**Comment by SleeplessByte** [comment]
Whilst this was fixed in the docs 11 months ago (thank you Petrik for getting that into the rails repo so quickly after this was merged :heart:), it seems that the fix wasn't pushed like until the first Rails 7 "docs" release, and it took over 6 months for this to make it to the non-edge guides. Happy to report that I've been (finally) able to use the docs again, for the past few months.

Whilst I no longer have a lot of respect for the "we are a thin team of volunteers who don't use windows"-argument that was given in the Rails repository, I want to, especially, thank everyone involved with this and related PRs. You may all be mouse-users yourself (and/or OSX), yet _all_ the issues I listed across multiple issues were fixed from your end in seven days. I don't know how to express the gratitude I feel, after I realised that I no longer was getting sad and frustrating whilst doing my jobs (and being in the docs).

Thank you!

**Comment by p8** [comment]
Thanks @SleeplessByte for the detailed response.
Much appreaciated! ❤️ 

Sadly, @MikeRogers0, who did most of the work, passed away last year 😢
